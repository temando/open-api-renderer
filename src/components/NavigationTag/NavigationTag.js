import React, { Component } from 'react'
import PropTypes from 'prop-types'

import Indicator from '../Indicator/Indicator'
import NavigationMethod from '../NavigationMethod/NavigationMethod'
import Description from '../Description/Description'

import './NavigationTag.scss'

export default class NavigationTag extends Component {
  constructor (props) {
    super(props)

    this.handleClick = this.handleClick.bind(this)
  }

  shouldComponentUpdate (nextProps, nextState) {
    const isHashDiff = this.props.location.hash !== nextProps.location.hash
    const isExpandedStatusDiff = this.props.shouldBeExpanded !== nextProps.shouldBeExpanded

    return isHashDiff || isExpandedStatusDiff
  }

  componentWillMount () {
    const { title, methods, location, onClick } = this.props

    if (!methods) {
      return null
    }

    const method = methods.find(
      (method) => (`#${method.link}` === location.hash)
    )

    if (method) {
      onClick(title)
    }
  }

  handleClick (e) {
    this.props.onClick(this.props.title)
  }

  render () {
    const { title, description, shouldBeExpanded, methods, location } = this.props

    // If tag has any method that matches location hash, then it is considered active
    let isActiveTag = false
    if (methods) {
      isActiveTag = methods.some(method => (`#${method.link}` === location.hash))
    }

    let isExpanded = false
    if (shouldBeExpanded || isActiveTag) {
      isExpanded = true
    }

    let indicatorDirection
    if (isExpanded) {
      indicatorDirection = 'bottom'
    } else {
      indicatorDirection = 'right'
    }

    return (
      <div>
        <a
          className='nav-tag'
          href={`#${title}`}
          onClick={this.handleClick}
        >
          <span>{title}</span>
          <Indicator direction={indicatorDirection} />
          {description && <Description description={description} />}
        </a>
        <div className='nav-tag-methods'>
          {methods && methods.map((method) => {
            const isActive = (`#${method.link}` === location.hash)

            return <NavigationMethod key={method.link} method={method} isActive={isActive} isOpen={isExpanded} />
          })}
        </div>
      </div>
    )
  }
}

NavigationTag.propTypes = {
  title: PropTypes.string.isRequired,
  description: PropTypes.string,
  methods: PropTypes.array,
  shouldBeExpanded: PropTypes.bool,
  onClick: PropTypes.func.isRequired,
  location: PropTypes.object.isRequired
}
