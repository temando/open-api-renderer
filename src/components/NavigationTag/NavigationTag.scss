.nav-tag {
  display: block;
  padding: 1rem 1rem .5rem;
  color: #FFF;

  &:hover {
    text-decoration: none;

    span {
      text-decoration: underline;
    }
  }

  &.active {
    background-color: rgb(12, 18, 29);
  }

  &.is-open {
    display: block;
  }

  &.is-closed {
    display: none;
  }
}
