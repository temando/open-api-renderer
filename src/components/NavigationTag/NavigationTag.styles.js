export const styles = theme => ({
  'navigationTagMethods': {
    '&$expanded': {
      marginBottom: '.5rem'
    }
  },

  'navigationTag': {
    display: 'block',
    padding: '1rem 1rem .5rem',

    '& span': {
      fontSize: `${theme.sizes.h4}`
    },

    '&:hover': {
      textDecoration: 'none',

      '& span': { textDecoration: 'underline' }
    }
  },

  'expanded': {}
})
