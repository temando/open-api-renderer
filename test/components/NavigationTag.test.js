import React from 'react'
import NavigationTag from './../../src/components/NavigationTag/NavigationTag'
import renderer from 'react-test-renderer'

describe('<NavigationTag />', () => {
  const location = {
    hash: '#pets'
  }

  it('renders with minimum props', () => {
    const tree = renderer.create(
      <NavigationTag title='pets' onClick={jest.fn()} location={location} />
    )

    expect(tree).toMatchSnapshot()
  })

  it('renders with methods', () => {
    const methods = [{
      type: 'get',
      title: 'Get pets',
      link: 'pets'
    },
    {
      type: 'get',
      title: 'Get pets',
      link: 'pets/:id'
    }]

    const tree = renderer.create(
      <NavigationTag title='pets' onClick={jest.fn()} location={location} methods={methods} />
    )

    expect(tree).toMatchSnapshot()
  })

  it('renders with description', () => {
    const tree = renderer.create(
      <NavigationTag title='Pets' handle='pets' description='Access to Pets' onClick={jest.fn()} location={location} />
    )

    expect(tree).toMatchSnapshot()
  })
})
