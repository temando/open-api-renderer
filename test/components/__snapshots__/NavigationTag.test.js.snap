// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NavigationTag /> renders with description 1`] = `
<div>
  <a
    className="nav-tag"
    href="#Pets"
    onClick={[Function]}
  >
    <span>
      Pets
    </span>
    <div
      className="indicator indicator--right"
    >
      <img
        alt=""
        src="test-file-stub"
        title="arrow"
      />
    </div>
    <div
      className="description"
      dangerouslySetInnerHTML={
        Object {
          "__html": "<p>Access to Pets</p>
        ",
        }
      }
    />
  </a>
  <div
    className="nav-tag-methods"
  />
</div>
`;

exports[`<NavigationTag /> renders with methods 1`] = `
<div>
  <a
    className="nav-tag"
    href="#pets"
    onClick={[Function]}
  >
    <span>
      pets
    </span>
    <div
      className="indicator indicator--bottom"
    >
      <img
        alt=""
        src="test-file-stub"
        title="arrow"
      />
    </div>
  </a>
  <div
    className="nav-tag-methods"
  >
    <a
      className="nav-method active open"
      href="#pets"
    >
      <span
        className="method-type"
      >
        GET
      </span>
      <span
        className="method-title"
      >
        Get pets
      </span>
    </a>
    <a
      className="nav-method open"
      href="#pets/:id"
    >
      <span
        className="method-type"
      >
        GET
      </span>
      <span
        className="method-title"
      >
        Get pets
      </span>
    </a>
  </div>
</div>
`;

exports[`<NavigationTag /> renders with minimum props 1`] = `
<div>
  <a
    className="nav-tag"
    href="#pets"
    onClick={[Function]}
  >
    <span>
      pets
    </span>
    <div
      className="indicator indicator--right"
    >
      <img
        alt=""
        src="test-file-stub"
        title="arrow"
      />
    </div>
  </a>
  <div
    className="nav-tag-methods"
  />
</div>
`;
